mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassin.java

```diff
@@ -30,32 +31,76 @@
 
 import com.github.fge.lambdas.Throwing;
 
-public class SpamAssassin {
+import reactor.core.Disposable;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.UnicastProcessor;
+import reactor.core.scheduler.Schedulers;
+import reactor.util.function.Tuple2;
+
+public class SpamAssassin implements Closeable {
 
     private final MetricFactory metricFactory;
-    private final SpamAssassinConfiguration spamAssassinConfiguration;
+    private Optional<UnicastProcessor<Tuple2<InputStream, Username>>> hamQueue;
+    private Optional<UnicastProcessor<Tuple2<InputStream, Username>>> spamQueue;
+    private Optional<Disposable> hamHandle;
+    private Optional<Disposable> spamHandle;
 
     @Inject
     public SpamAssassin(MetricFactory metricFactory, SpamAssassinConfiguration spamAssassinConfiguration) {
         this.metricFactory = metricFactory;
-        this.spamAssassinConfiguration = spamAssassinConfiguration;
+
+        createListeners(spamAssassinConfiguration);
     }
 
-    public void learnSpam(List<InputStream> messages, Username username) {
-        if (spamAssassinConfiguration.isEnable()) {
-            Host host = spamAssassinConfiguration.getHost().get();
-            SpamAssassinInvoker invoker = new SpamAssassinInvoker(metricFactory, host.getHostName(), host.getPort());
-            messages
-                .forEach(Throwing.consumer(message -> invoker.learnAsSpam(message, username)));
-        }
+    public void learnSpam(Flux<InputStream> messages, Username username) {
+        spamQueue.ifPresent(queue ->
+            messages.zipWith(Mono.just(username).repeat())
+                .doOnNext(queue::onNext)
+                .then()
+                .block()
+        );
     }
 
-    public void learnHam(List<InputStream> messages, Username username) {
+    public void learnHam(Flux<InputStream> messages, Username username) {
+        hamQueue.ifPresent(queue ->
+            messages.zipWith(Mono.just(username).repeat())
+                .doOnNext(queue::onNext)
+                .then()
+                .block()
+        );
+    }
+
+    @Override
+    public void close() {
+        hamHandle.ifPresent(Disposable::dispose);
+        spamHandle.ifPresent(Disposable::dispose);
+    }
+
+    private void createListeners(SpamAssassinConfiguration spamAssassinConfiguration) {
         if (spamAssassinConfiguration.isEnable()) {
             Host host = spamAssassinConfiguration.getHost().get();
             SpamAssassinInvoker invoker = new SpamAssassinInvoker(metricFactory, host.getHostName(), host.getPort());
-            messages
-                .forEach(Throwing.consumer(message -> invoker.learnAsHam(message, username)));
+
+            UnicastProcessor<Tuple2<InputStream, Username>> ham = UnicastProcessor.create();
```

Comment:
> A Processor implementation that takes a custom queue and allows only a single subscriber. 

This makes the assumption that we won't gain anything from paralelizing calls to spamassassin.

What backs this assumption?

Reply:
What's the overflow strategy? will we have good enough back-pressure?

Reply:
> This makes the assumption that we won't gain anything from paralelizing calls to spamassassin.
>
> What backs this assumption?

SA was slowing the delivery in production. Anyway, we could define a configurable parallelism.

> What's the overflow strategy? will we have good enough back-pressure?

Not for the moment, but we can discuss it

Reply:
The best overflow strategy would be to size the queue and not dequeue faster from the EventBus than what this class can handle. Buffer queue size have to be sized on that IMO

Reply:
This strategy is irrelevant given https://github.com/linagora/james-project/pull/3253 ?

Reply:
No #3253 is partially irrelevant thanks to my PR